core/src/transport.rs

```diff
@@ -129,17 +140,27 @@ pub trait Transport {
     ///
     /// If [`TransportError::MultiaddrNotSupported`] is returned, it may be desirable to
     /// try an alternative [`Transport`], if available.
-    fn dial(&mut self, addr: Multiaddr) -> Result<Self::Dial, TransportError<Self::Error>>;
+    fn dial(
+        &mut self,
+        addr: Multiaddr,
+        opts: DialOpts,
+    ) -> Result<Self::Dial, TransportError<Self::Error>>;
```

Comment:
We should think about is whether we _want_ to design this interface in a forward compatible way, i.e. to allow us to make changes without them being breaking.

To do that, we'd have to:

- Make all fields of `DialOpts` private
- Provide constructors that `libp2p-swarm` can use
- Provide boolean(?) accessors that the various transports can use

I'd say we either go down this route - or - if we don't care about forward compatibility, we remove the `DialOpts` struct and make two arguments. That is a simpler interface to understand but ties us down more.

I think to start with, I'd prefer the simpler interface. We can pick up this discussion again once we have all the things in place :)

Reply:
We can go either way, although I honestly prefer DialOpts, primarily because there a lot of instances where we can just pass the option down.

However can you explain to me why we can't just keep the fields public? I don't see us doing a lot of changes and it would make writing code more fun (destructuring).

Reply:
> However can you explain to me why we can't just keep the fields public? I don't see us doing a lot of changes and it would make writing code more fun (destructuring). If you are concerned about making breaking changes by adding fields, we could just use non_exhaustive

Hmm that is fair I guess. Let's go with `DialOpts` then!